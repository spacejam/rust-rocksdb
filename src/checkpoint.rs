// Copyright 2018 Eugene P.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

//! Implementation of bindings to RocksDB Checkpoint[1] API
//!
//! [1]: https://github.com/facebook/rocksdb/wiki/Checkpoints

use crate::{ffi, ops::CheckpointInternal, Error};
use std::ffi::CString;
use std::marker::PhantomData;
use std::path::Path;

/// Undocumented parameter for `ffi::rocksdb_checkpoint_create` function. Zero by default.
const LOG_SIZE_FOR_FLUSH: u64 = 0_u64;

/// Database's checkpoint object.
/// Used to create checkpoints of the specified DB from time to time.
pub struct Checkpoint<'a> {
    inner: *mut ffi::rocksdb_checkpoint_t,
    _db: PhantomData<&'a ()>,
}

impl<'a> Checkpoint<'a> {
    /// Creates new checkpoint object for specific DB.
    ///
    /// Does not actually produce checkpoints, call `.create_checkpoint()` method to produce
    /// a DB checkpoint.
    pub fn new<C: CheckpointInternal>(db: &'a C) -> Result<Checkpoint<'a>, Error> {
        let checkpoint: *mut ffi::rocksdb_checkpoint_t = unsafe { db.create_checkpoint_object()? };

        if checkpoint.is_null() {
            return Err(Error::new("Could not create checkpoint object.".to_owned()));
        }

        Ok(Checkpoint {
            inner: checkpoint,
            _db: PhantomData,
        })
    }

    /// Creates new physical DB checkpoint in directory specified by `path`.
    pub fn create_checkpoint<P: AsRef<Path>>(&self, path: P) -> Result<(), Error> {
        let path = path.as_ref();
        let cpath = if let Ok(c) = CString::new(path.to_string_lossy().as_bytes()) {
            c
        } else {
            return Err(Error::new(
                "Failed to convert path to CString when creating DB checkpoint".to_owned(),
            ));
        };

        unsafe {
            ffi_try!(ffi::rocksdb_checkpoint_create(
                self.inner,
                cpath.as_ptr(),
                LOG_SIZE_FOR_FLUSH,
            ));

            Ok(())
        }
    }
}

impl<'a> Drop for Checkpoint<'a> {
    fn drop(&mut self) {
        unsafe {
            ffi::rocksdb_checkpoint_object_destroy(self.inner);
        }
    }
}
