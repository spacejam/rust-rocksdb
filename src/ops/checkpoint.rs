// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

use crate::{db::DBInner, ffi, handle::Handle, Error};
use ambassador::delegatable_trait;

/// This is an internal trait used to create and free a checkpoint
#[delegatable_trait]
pub trait CheckpointInternal {
    unsafe fn create_checkpoint_object(&self) -> Result<*mut ffi::rocksdb_checkpoint_t, Error>;
}

impl CheckpointInternal for DBInner {
    unsafe fn create_checkpoint_object(&self) -> Result<*mut ffi::rocksdb_checkpoint_t, Error> {
        Ok(ffi_try!(ffi::rocksdb_checkpoint_object_create(
            self.handle()
        )))
    }
}
