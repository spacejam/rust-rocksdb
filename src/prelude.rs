pub use crate::{ops::*, ColumnFamily, DBVector, Error, Options, ReadOptions, WriteOptions, DB};
