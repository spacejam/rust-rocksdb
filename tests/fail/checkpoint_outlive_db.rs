use rocksdb::prelude::*;
use rocksdb::checkpoint::Checkpoint;

fn main() {
    let _checkpoint = {
        let db = DB::open_default("foo").unwrap();
        Checkpoint::new(&db)
    };
}
