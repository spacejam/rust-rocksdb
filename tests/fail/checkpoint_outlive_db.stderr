error[E0597]: `db` does not live long enough
 --> $DIR/checkpoint_outlive_db.rs:7:25
  |
5 |     let _checkpoint = {
  |         ----------- borrow later stored here
6 |         let db = DB::open_default("foo").unwrap();
7 |         Checkpoint::new(&db)
  |                         ^^^ borrowed value does not live long enough
8 |     };
  |     - `db` dropped here while still borrowed
